src/local/room_coordinate.rs

```diff
@@ -0,0 +1,136 @@
+use std::convert::TryFrom;
+use std::fmt;
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
+
+pub const ROOM_SIZE: u8 = 50;
+pub const ROOM_AREA: usize = (ROOM_SIZE as usize) * (ROOM_SIZE as usize);
+
+#[derive(Debug, Clone, Copy)]
+pub struct OutOfBoundsError(u8);
+
+impl fmt::Display for OutOfBoundsError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Out of bounds coordinate: {}", self.0)
+    }
+}
+
+#[inline]
+pub fn xy_to_linear_index(xy: RoomXY) -> usize {
+    ((xy.x.0 as usize) * (ROOM_SIZE as usize)) + (xy.y.0 as usize)
+}
+
+#[inline]
+pub fn linear_index_to_xy(idx: usize) -> RoomXY {
+    assert!(idx < ROOM_AREA, "Out of bounds index: {}", idx);
+    // SAFETY: bounds checking above ensures both are within range.
+    RoomXY {
+        x: unsafe { RoomCoordinate::unchecked_new((idx / (ROOM_SIZE as usize)) as u8) },
+        y: unsafe { RoomCoordinate::unchecked_new((idx % (ROOM_SIZE as usize)) as u8) }
+    }
+}
+
+#[derive(Debug, Hash, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RoomCoordinate(u8);
+
+#[derive(Debug, Hash, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RoomXY {
+    pub x: RoomCoordinate,
+    pub y: RoomCoordinate
+}
+
+impl fmt::Display for RoomXY {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "({}, {})", self.x, self.y)
+    }
+}
+
+impl RoomCoordinate {
+    // # Safety
+    // Calling this method with `coord >= 50` can result in undefined behaviour when used.
+    #[inline]
+    pub unsafe fn unchecked_new(coord: u8) -> Self {
+        debug_assert!(coord < ROOM_SIZE, "Out of bounds unchecked coordinate: {}", coord);
+        RoomCoordinate(coord)
+    }
+
+    #[inline]
+    pub fn val(self) -> u8 {
```

Comment:
I would recommend implementing Deref or just using into() which is more standard. Looking at the rest of the code 'val()' becomes very repetitive - using more idiomatic rust make it more obvious. Additionally you could get rid of some of the other casts where the 'into' output type could be inferred.

Reply:
It has the From<RoomCoordinate> for u8, so sure, this becomes redundant.